File: src/merkle/getMerkleRootFromProof.ts
// Copyright (C) 2022 Deliberative Technologies P.C.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// 	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import dcryptoMemory from "./memory";

import dcryptoMethodsModule from "../c/build/dcryptoMethodsModule";

import { crypto_hash_sha512_BYTES } from "../utils/interfaces";

/**
 * @function
 * Calculates the Merkle root from the element hash and its Merkle proof.
 *
 * @param hash: The hash of the base element in question.
 * @param proof: The first element is the first leave that was added for the calculation etc. The last
 * byte is either 0 or 1, indicating whether it is to the left or to the right in the tree.
 *
 * @returns The Merkle root
 */
const getMerkleRootFromProof = async (
  hash: Uint8Array,
  proof: Uint8Array,
): Promise<Uint8Array> => {
  const proofLen = proof.length;
  if (proofLen % (crypto_hash_sha512_BYTES + 1) !== 0)
    throw new Error("Proof length not multiple of hash length + 1.");
  const proofArtifactsLen = proofLen / (crypto_hash_sha512_BYTES + 1);

  const wasmMemory = dcryptoMemory.verifyMerkleProofMemory(proofLen);
  const module = await dcryptoMethodsModule({
    wasmMemory,
  });

  const ptr1 = module._malloc(crypto_hash_sha512_BYTES);
  const elementHash = new Uint8Array(
    module.HEAPU8.buffer,
    ptr1,
    crypto_hash_sha512_BYTES,
  );
  elementHash.set(hash);

  const ptr2 = module._malloc(proofLen);
  const proofArray = new Uint8Array(module.HEAPU8.buffer, ptr2, proofLen);
  proofArray.set(proof);

  const ptr3 = module._malloc(crypto_hash_sha512_BYTES);
  const rootArray = new Uint8Array(
    module.HEAPU8.buffer,
    ptr3,
    crypto_hash_sha512_BYTES,
  );

  const result = module._get_merkle_root_from_proof(
    proofArtifactsLen,
    elementHash.byteOffset,
    proofArray.byteOffset,
    rootArray.byteOffset,
  );

  module._free(ptr1);
  module._free(ptr2);

  switch (result) {
    case 0: {
      const proof = Uint8Array.from(rootArray);
      module._free(ptr3);

      return proof;
    }

    case -1: {
      module._free(ptr3);

      throw new Error("Proof artifact position is neither left nor right.");
    }

    case -2: {
      module._free(ptr3);

      throw new Error("Could not calculate hash.");
    }

    default: {
      module._free(ptr3);

      throw new Error("Unexpected error occured.");
    }
  }
};

export default getMerkleRootFromProof;

File: src/c/merkle/get_merkle_root_from_proof.c
// Copyright (C) 2022 Deliberative Technologies P.C.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// 	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <stdbool.h>
#include <stdint.h>
#include <string.h>

#include "../../../libsodium/src/libsodium/include/sodium/crypto_hash_sha512.h"

__attribute__((used)) int
get_merkle_root_from_proof(
    const unsigned int PROOF_ARTIFACTS_LEN,
    const uint8_t element_hash[crypto_hash_sha512_BYTES],
    const uint8_t proof[PROOF_ARTIFACTS_LEN][crypto_hash_sha512_BYTES + 1],
    uint8_t root[crypto_hash_sha512_BYTES])
{
  memcpy(root, element_hash, crypto_hash_sha512_BYTES);

  size_t i;
  unsigned int position;
  int res;

  if (PROOF_ARTIFACTS_LEN == 1)
  {
    bool isOne = true;
    for (i = 0; i < crypto_hash_sha512_BYTES + 1; i++)
    {
      if (proof[0][i] != 1)
      {
        isOne = false;
        break;
      }
    }

    // Single element tree
    if (isOne) return 0;
  }

  uint8_t(*concat_hashes)[crypto_hash_sha512_BYTES]
      = malloc(sizeof(uint8_t[2][crypto_hash_sha512_BYTES]));
  if (concat_hashes == NULL) return -1;

  for (i = 0; i < PROOF_ARTIFACTS_LEN; i++)
  {
    position = proof[i][crypto_hash_sha512_BYTES];

    // Proof artifact goes to the left
    if (position == 0)
    {
      memcpy(&concat_hashes[0], &proof[i][0], crypto_hash_sha512_BYTES);
      memcpy(&concat_hashes[1], &root[0], crypto_hash_sha512_BYTES);
    }
    else if (position == 1)
    {
      memcpy(&concat_hashes[0], &root[0], crypto_hash_sha512_BYTES);
      memcpy(&concat_hashes[1], &proof[i][0], crypto_hash_sha512_BYTES);
    }
    else
    {
      free(concat_hashes);

      return -2;
    }

    res = crypto_hash_sha512(root, *concat_hashes,
                             2 * crypto_hash_sha512_BYTES);
    if (res != 0)
    {
      free(concat_hashes);

      return -3;
    }
  }

  free(concat_hashes);

  return 0;
}
